### PR TITLE
Allow servers to select compression level via initial MD & overall compression cleanup

### PR DESCRIPTION
**Pull request by dgquintas**
Setting the newly added `compression_level` field of `grpc_op::send_initial_metadata` _by a server_ now has the effect of applying that compression level for the subsequent call messages leaving the server. The ultimate meaning of the level depends on the client's supported compression algorithms.

### CODE REVIEW

**Comment by murgatroid99** [comment]
I thought the point of adding `memset` everywhere was to `memset` the whole op to 0 when it is created, so that we would never have to do this again.

**Comment by ctiller** [comment]
That was the intent.

**Comment by dgquintas** [comment]
Yes, agreed. Unintentional omission (current state was just easier to grep for as a starting point). Fixing.

In any case the most review-worthy part of the code is the changes to `surface/call.c`.

**Comment by jtattermusch** [comment]
It seems like using the default value of CompressionLevel=None as a default for the send initial metadata operation is problematic if you also allow setting compression level/algorithm at the channel or server level.

e.g. If you set a channel-wide setting of default level being "LOW" let's say,   does   CompressionLevel=None in the send metadata operation means you want to disable compression for that call or that you wanna go with the default channel-wide setting? "LOW" here.
Looks like we would be better of with `data.send_initial_metadata.compression_level`  being nullable.

**Comment by dgquintas** [comment]
@jtattermusch addressesed on latest commit f210b43. PTAL.

**Comment by jtattermusch** [comment]
The fix itself looks fine to me.

There's a build error:
```
[CXX]     Compiling src/cpp/client/client_context.cc
In file included from include/grpc++/impl/call.h:37:0,
                 from include/grpc++/channel.h:39,
                 from src/cpp/client/channel.cc:34:
include/grpc++/impl/codegen/call.h: In member function 'void grpc::CallOpSendInitialMetadata::AddOp(grpc_op*, size_t*)':
include/grpc++/impl/codegen/call.h:205:36: error: 'struct grpc_op::<anonymous union>::<anonymous>' has no member named 'compression_level'
     op->data.send_initial_metadata.compression_level = compression_level_;
                                    ^
Makefile:2126: recipe for target '/var/local/git/grpc/objs/dbg/src/cpp/client/channel.o' failed
make: *** [/var/local/git/grpc/objs/dbg/src/cpp/client/channel.o] Error 1
make: *** Waiting for unfinished jobs....
Command exited with non-zero status 2
real 15.01
user 18.24
sys 3.12

FAILED: make [ret=2, pid=590]
```

**Comment by dgquintas** [comment]
@jtattermusch Fixed.

File: include/grpc/impl/codegen/grpc_types.h [hunk 1]
```diff
@@ -332,7 +333,12 @@ typedef struct grpc_op {
     struct {
       size_t count;
       grpc_metadata *metadata;
-      grpc_compression_level compression_level;
+      /** If \a is_set, \a compression_level will be used for the call.
+       * Otherwise, \a compression_level won't be considered */
+      struct {
+        bool is_set;
+        grpc_compression_level compression_level;
+      } maybe_compression_level;
```
Comment (dgquintas) [inline include/grpc/impl/codegen/grpc_types.h]: @ctiller do you think it'd be worth it to encode the `is_set` bit within `compression_level` itself with some bit twiddling as we already do in other places? I don't think the small size increase of the op is important, given how it's translated into a stream_op anyway...

Comment (ctiller) [inline include/grpc/impl/codegen/grpc_types.h]: We reserve 8 pointers for op data (see a few lines above): we are fine here.


That said, this is a public header and must be c89. Please lose the bool.

Comment (dgquintas) [inline include/grpc/impl/codegen/grpc_types.h]: Ah, I see, thanks for pointing that out.

Done.

File: src/core/lib/surface/call.c [hunk 2]
```diff
@@ -1229,15 +1319,40 @@ static grpc_call_error call_start_batch(grpc_exec_ctx *exec_ctx,
...
+            error = GRPC_CALL_ERROR_NOT_ON_CLIENT;
+            goto done_with_error;
+          }
+          const grpc_compression_algorithm calgo =
+              compression_algorithm_for_level_locked(
+                  call, op->data.send_initial_metadata.maybe_compression_level
+                            .compression_level);
+          char *calgo_name;
+          grpc_compression_algorithm_name(calgo, &calgo_name);
+          compression_md.key = "grpc-internal-encoding-request";
```
Comment (dgquintas) [inline src/core/lib/surface/call.c]: Note to self: use define from compression_types.h instead

**Comment by dgquintas** [comment]
Failure is flaky

**Comment by dgquintas** [comment]
Ping :)

**Comment by dgquintas** [comment]
Fixes #6219

**Comment by sreecha** [comment]
Had one minor comment which is fixed in the latest commit. 

LGTM!..

**Comment by dgquintas** [comment]
test this please

**Comment by dgquintas** [comment]
Known flakes: run_ruby and openloop

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* One or more commits are not associated with a GitHub user. The email used in the Git commit must match a verified email address in a GitHub user profile. For more information, please see https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/

_Regards,
CLA GitHub bot_

**Pull request merged**